src/common/transformations/src/transformations/mlir/mlir_op.cpp

```diff
@@ -267,6 +290,138 @@ namespace mlir {
 
 using namespace ::mlir;
 
+std::shared_ptr<MLIREvaluateBase> MLIREvaluateBase::create(OwningOpRef<ModuleOp> module,
+                                                           MlirMode mode,
+                                                           std::shared_ptr<ov::EvaluationContext> loweringContext) {
+    switch (mode) {
+        #ifdef GC_USE_IMEX // GC_GPU requires IMEX support
+        case MLIR_MODE_GC_GPU:
+            return std::make_shared<MLIREvaluateGcGPU>(std::move(module), loweringContext);
+        #endif
+        case MLIR_MODE_TPP:
+        case MLIR_MODE_GC:
+        case MLIR_MODE_DEFAULT:
+            return std::make_shared<MLIREvaluate>(std::move(module), mode);
+        default:
+            OPENVINO_THROW("Unsupported MLIR mode");
+    }
+}
+
+#ifdef GC_USE_IMEX // GC_GPU requires IMEX support
+
+cl_device_id extract_device_from_context(cl_context context) {
+    size_t devices_size;
+    cl_int err = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, NULL, &devices_size);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting context info: ", err);
+    }
+    if (devices_size / sizeof(cl_device_id) != 1) {
+        OPENVINO_THROW("Expected exactly one device in the context, got ", devices_size);
+    }
+
+    cl_device_id devices;
+    err = clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size, &devices, NULL);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting device IDs: ", err);
+    }
+
+    return devices;
+}
+
+
+
+MLIREvaluateGcGPU::MLIREvaluateGcGPU(OwningOpRef<mlir::ModuleOp> _module, std::shared_ptr<ov::EvaluationContext> loweringContext) {
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "[ DEBUG ] Source MLIR:\n"
+        "-----------------------------------------\n");
+    OPENVINO_MLIR_DEBUG(_module->dump());
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "-----------------------------------------\n");
+
+    gc::gpu::OclModuleBuilder builder(std::move(_module));
+
+    auto it = loweringContext->find(ov::intel_gpu::ocl_context.name());
+    if (it == loweringContext->end()) {
+        OPENVINO_THROW("No cl_context provided for OpenCL execution");
+    }
+    // QUESTION: are we sure that there's always only one device per context?
+    // QUESTION: may we pass context as void* directly to 'builder.build()' and make it responsible
+    // for extracting the device?
+    auto context = reinterpret_cast<cl_context>(it->second.as<ov::intel_gpu::gpu_handle_param>());
+    module = gcGetOrReport(builder.build(extract_device_from_context(context), context));
```

Comment:
```suggestion
    if (auto mod = builder.build(extract_device_from_context(context), context)) {
        module = *mod;
    } else {
        OPENVINO_THROW("Failed to build OclModule");
    }
```

Probably, OPENVINO_THROW() should be used here.

Reply:
Or we can add an option to the GPU pipeline. It could be done in a separate PR.